File: implementations/elixir/lib/rhizomatic/pack.ex
defmodule Rhizomatic.Pack do
  @moduledoc """
  The L0 pack (SPEC-8): a physical container for a delta set, one canonical
  CBOR item in the Rhizomatic profile. Determinism is total: same delta set
  ⇒ same pack bytes ⇒ same packId.

  Layout (SPEC-8 §3, plus the `"i"` id field the §4 rehydration contract
  presupposes — see FINDINGS.md F2–F4):

      Pack = { "version": 1, "strings": [tstr...],
               "envelopes": [Record...], "members": [MemberRecord...],
               "loose": [Record...] }

      Record       = { "a": authorIdx, "i": idIdx, "t": timestamp,
                       "p": [Ptr...], "s"?: sigIdx }
      MemberRecord = { "m": envelopeIdx, "i": idIdx, "p": [Ptr...],
                       "a"?: authorIdx, "dt"?: number, "s"?: sigIdx }
      Ptr          = { "r": roleIdx,
                       "e"|"d"|"s": idx | "n": number | "b": bool
                                  | ("m": mimeIdx, "y": bstr),
                       "c"?: ctxIdx }

  All indices are positions in the sorted-unique `strings` table, encoded in
  the profile's float form. Unpacking rehydrates every record and MUST
  verify its content address against the stored id (fsck for free).
  """

  alias Rhizomatic.{Cbor, Delta, Hash}

  @txn_member "rhizomatic.txn.member"

  @typedoc "A delta as packed/unpacked: validated claims plus optional hex sig."
  @type delta :: %{claims: Delta.claims(), sig: String.t() | nil}

  # ------------------------------------------------------------------- pack

  @spec pack([delta()]) :: %{bytes: binary(), pack_id: String.t()}
  def pack(deltas) do
    entries =
      deltas
      |> Enum.map(fn d ->
        %{claims: d.claims, sig: Map.get(d, :sig), id: Delta.id_hex(d.claims)}
      end)
      |> Enum.uniq_by(& &1.id)

    envelopes = entries |> Enum.filter(&manifest?/1) |> Enum.sort_by(& &1.id)
    envelope_ids = MapSet.new(envelopes, & &1.id)

    # member id -> lexicographically-first claiming manifest (by manifest id)
    claimed =
      envelopes
      |> Enum.flat_map(fn m -> Enum.map(member_ids(m), &{&1, m.id}) end)
      |> Enum.reduce(%{}, fn {member_id, manifest_id}, acc ->
        Map.update(acc, member_id, manifest_id, &min(&1, manifest_id))
      end)

    members =
      entries
      |> Enum.filter(fn e -> Map.has_key?(claimed, e.id) and e.id not in envelope_ids end)
      |> Enum.sort_by(& &1.id)

    loose =
      entries
      |> Enum.filter(fn e -> not Map.has_key?(claimed, e.id) and e.id not in envelope_ids end)
      |> Enum.sort_by(& &1.id)

    strings = string_table(entries)
    idx = strings |> Enum.with_index() |> Map.new()
    envelope_index = envelopes |> Enum.map(& &1.id) |> Enum.with_index() |> Map.new()
    envelope_by_id = Map.new(envelopes, &{&1.id, &1})

    pack_ast =
      {:map,
       [
         {{:tstr, "version"}, {:float, 1.0}},
         {{:tstr, "strings"}, {:arr, Enum.map(strings, &{:tstr, &1})}},
         {{:tstr, "envelopes"}, {:arr, Enum.map(envelopes, &record_ast(&1, idx))}},
         {{:tstr, "members"},
          {:arr,
           Enum.map(members, fn e ->
             manifest = envelope_by_id[claimed[e.id]]
             member_ast(e, manifest, envelope_index[manifest.id], idx)
           end)}},
         {{:tstr, "loose"}, {:arr, Enum.map(loose, &record_ast(&1, idx))}}
       ]}

    bytes = Cbor.encode(pack_ast)
    %{bytes: bytes, pack_id: Hash.id_hex(bytes)}
  end

  defp manifest?(%{claims: %{pointers: pointers}}) do
    Enum.any?(pointers, fn
      %{role: @txn_member, target: {:delta, _, _}} -> true
      _ -> false
    end)
  end

  defp member_ids(%{claims: %{pointers: pointers}}) do
    for %{role: @txn_member, target: {:delta, hex, _}} <- pointers, do: hex
  end

  # every string in the set interns: delta ids, sig hexes, authors, roles,
  # contexts, entity ids, delta-ref hexes, string primitives, mimes —
  # sorted unique, bytewise
  defp string_table(entries) do
    entries
    |> Enum.flat_map(fn e ->
      [e.id, e.claims.author] ++
        if(e.sig, do: [e.sig], else: []) ++
        Enum.flat_map(e.claims.pointers, fn %{role: role, target: t} ->
          [role | target_strings(t)]
        end)
    end)
    |> Enum.uniq()
    |> Enum.sort()
  end

  defp target_strings({:string, s}), do: [s]
  defp target_strings({:number, _}), do: []
  defp target_strings({:boolean, _}), do: []
  defp target_strings({:entity, id, ctx}), do: [id | if(ctx, do: [ctx], else: [])]
  defp target_strings({:delta, hex, ctx}), do: [hex | if(ctx, do: [ctx], else: [])]
  defp target_strings({:bytes, mime, _}), do: [mime]

  # hydrated Record for envelopes and loose deltas
  defp record_ast(e, idx) do
    {:map,
     [
       {{:tstr, "a"}, fidx(idx, e.claims.author)},
       {{:tstr, "i"}, fidx(idx, e.id)},
       {{:tstr, "t"}, {:float, e.claims.timestamp}},
       {{:tstr, "p"}, {:arr, Enum.map(e.claims.pointers, &ptr_ast(&1, idx))}}
     ] ++ sig_pair(e, idx)}
  end

  # dehydrated MemberRecord relative to its claiming manifest
  defp member_ast(e, manifest, envelope_idx, idx) do
    dt = e.claims.timestamp - manifest.claims.timestamp

    {:map,
     [
       {{:tstr, "m"}, {:float, envelope_idx * 1.0}},
       {{:tstr, "i"}, fidx(idx, e.id)},
       {{:tstr, "p"}, {:arr, Enum.map(e.claims.pointers, &ptr_ast(&1, idx))}}
     ] ++
       if(e.claims.author == manifest.claims.author,
         do: [],
         else: [{{:tstr, "a"}, fidx(idx, e.claims.author)}]
       ) ++
       if(dt == 0.0, do: [], else: [{{:tstr, "dt"}, {:float, dt}}]) ++
       sig_pair(e, idx)}
  end

  defp sig_pair(%{sig: nil}, _idx), do: []
  defp sig_pair(%{sig: sig}, idx), do: [{{:tstr, "s"}, fidx(idx, sig)}]

  defp ptr_ast(%{role: role, target: target}, idx) do
    {:map, [{{:tstr, "r"}, fidx(idx, role)} | target_ptr_pairs(target, idx)]}
  end

  defp target_ptr_pairs({:string, s}, idx), do: [{{:tstr, "s"}, fidx(idx, s)}]
  defp target_ptr_pairs({:number, f}, _idx), do: [{{:tstr, "n"}, {:float, f}}]
  defp target_ptr_pairs({:boolean, b}, _idx), do: [{{:tstr, "b"}, {:bool, b}}]

  defp target_ptr_pairs({:entity, id, ctx}, idx),
    do: [{{:tstr, "e"}, fidx(idx, id)} | ctx_pair(ctx, idx)]

  defp target_ptr_pairs({:delta, hex, ctx}, idx),
    do: [{{:tstr, "d"}, fidx(idx, hex)} | ctx_pair(ctx, idx)]

  defp target_ptr_pairs({:bytes, mime, payload}, idx),
    do: [{{:tstr, "m"}, fidx(idx, mime)}, {{:tstr, "y"}, {:bstr, payload}}]

  defp ctx_pair(nil, _idx), do: []
  defp ctx_pair(ctx, idx), do: [{{:tstr, "c"}, fidx(idx, ctx)}]

  defp fidx(idx, s), do: {:float, Map.fetch!(idx, s) * 1.0}

  # ----------------------------------------------------------------- unpack

  @doc """
  Decode pack bytes back into the delta set. Every record is rehydrated
  through the standard delta-construction path and its content address MUST
  equal the stored id — a record that does not rehydrate fails the unpack.
  """
  @spec unpack(binary()) :: {:ok, [delta()]} | {:error, term()}
  def unpack(bytes) do
    with {:ok, {:map, pairs}} <- Cbor.decode_exact(bytes),
         {:ok, m} <- pack_fields(pairs),
         {:ok, strings} <- unpack_strings(m["strings"]),
         {:ok, envelopes} <- unpack_records(m["envelopes"], strings),
         {:ok, loose} <- unpack_records(m["loose"], strings),
         {:ok, members} <- unpack_members(m["members"], strings, envelopes) do
      all = envelopes ++ members ++ loose

      Enum.reduce_while(all, {:ok, []}, fn {delta, stored_id}, {:ok, acc} ->
        with {:ok, claims} <- Delta.validate(delta.claims),
             true <- Delta.id_hex(claims) == stored_id || {:error, {:id_mismatch, stored_id}} do
          {:cont, {:ok, [%{claims: claims, sig: delta.sig} | acc]}}
        else
          {:error, e} -> {:halt, {:error, e}}
        end
      end)
      |> case do
        {:ok, acc} -> {:ok, Enum.reverse(acc)}
        err -> err
      end
    else
      {:error, e} -> {:error, e}
      _ -> {:error, :malformed_pack}
    end
  end

  defp pack_fields(pairs) do
    m =
      Map.new(pairs, fn
        {{:tstr, k}, v} -> {k, v}
        {_, v} -> {:bad_key, v}
      end)

    with {:float, 1.0} <- Map.get(m, "version", :missing) do
      if Enum.sort(Map.keys(m)) == ["envelopes", "loose", "members", "strings", "version"] do
        {:ok, m}
      else
        {:error, :malformed_pack}
      end
    else
      _ -> {:error, :unsupported_pack_version}
    end
  end

  defp unpack_strings({:arr, items}) do
    Enum.reduce_while(items, {:ok, []}, fn
      {:tstr, s}, {:ok, acc} -> {:cont, {:ok, [s | acc]}}
      _, _ -> {:halt, {:error, :malformed_strings}}
    end)
    |> case do
      {:ok, acc} -> {:ok, acc |> Enum.reverse() |> List.to_tuple()}
      err -> err
    end
  end

  defp unpack_strings(_), do: {:error, :malformed_strings}

  defp unpack_records({:arr, records}, strings) do
    map_ok(records, fn {:map, pairs} ->
      m = field_map(pairs)

      with {:ok, author} <- table(strings, m["a"]),
           {:ok, id} <- table(strings, m["i"]),
           {:float, ts} <- Map.get(m, "t", :missing),
           {:ok, pointers} <- unpack_ptrs(m["p"], strings),
           {:ok, sig} <- opt_table(strings, m["s"]) do
        {:ok, {%{claims: %{timestamp: ts, author: author, pointers: pointers}, sig: sig}, id}}
      else
        _ -> {:error, :malformed_record}
      end
    end)
  end

  defp unpack_records(_, _), do: {:error, :malformed_pack}

  defp unpack_members({:arr, records}, strings, envelopes) do
    env = List.to_tuple(envelopes)

    map_ok(records, fn {:map, pairs} ->
      m = field_map(pairs)

      with {:ok, env_idx} <- as_index(m["m"]),
           true <- env_idx < tuple_size(env) || {:error, :envelope_index_out_of_range},
           {manifest, _} <- elem(env, env_idx),
           {:ok, id} <- table(strings, m["i"]),
           {:ok, pointers} <- unpack_ptrs(m["p"], strings),
           {:ok, author} <-
             (case m["a"] do
                nil -> {:ok, manifest.claims.author}
                v -> table(strings, v)
              end),
           {:ok, ts} <-
             (case m["dt"] do
                nil -> {:ok, manifest.claims.timestamp}
                {:float, dt} -> {:ok, manifest.claims.timestamp + dt}
                _ -> {:error, :malformed_record}
              end),
           {:ok, sig} <- opt_table(strings, m["s"]) do
        {:ok, {%{claims: %{timestamp: ts, author: author, pointers: pointers}, sig: sig}, id}}
      else
        _ -> {:error, :malformed_record}
      end
    end)
  end

  defp unpack_members(_, _, _), do: {:error, :malformed_pack}

  defp unpack_ptrs({:arr, ptrs}, strings) do
    map_ok(ptrs, fn {:map, pairs} ->
      m = field_map(pairs)

      with {:ok, role} <- table(strings, m["r"]),
           {:ok, target} <- unpack_target(m, strings) do
        {:ok, %{role: role, target: target}}
      else
        _ -> {:error, :malformed_ptr}
      end
    end)
  end

  defp unpack_ptrs(_, _), do: {:error, :malformed_pack}

  defp unpack_target(m, strings) do
    ctx =
      case m["c"] do
        nil -> {:ok, nil}
        v -> table(strings, v)
      end

    with {:ok, context} <- ctx do
      cond do
        Map.has_key?(m, "e") ->
          with {:ok, id} <- table(strings, m["e"]), do: {:ok, {:entity, id, context}}

        Map.has_key?(m, "d") ->
          with {:ok, hex} <- table(strings, m["d"]), do: {:ok, {:delta, hex, context}}

        Map.has_key?(m, "s") ->
          with {:ok, s} <- table(strings, m["s"]), do: {:ok, {:string, s}}

        Map.has_key?(m, "n") ->
          case m["n"] do
            {:float, f} -> {:ok, {:number, f}}
            _ -> {:error, :malformed_ptr}
          end

        Map.has_key?(m, "b") ->
          case m["b"] do
            {:bool, b} -> {:ok, {:boolean, b}}
            _ -> {:error, :malformed_ptr}
          end

        Map.has_key?(m, "y") ->
          with {:ok, mime} <- table(strings, m["m"]),
               {:bstr, payload} <- m["y"] do
            {:ok, {:bytes, mime, payload}}
          else
            _ -> {:error, :malformed_ptr}
          end

        true ->
          {:error, :malformed_ptr}
      end
    end
  end

  defp field_map(pairs) do
    Map.new(pairs, fn
      {{:tstr, k}, v} -> {k, v}
      {k, v} -> {k, v}
    end)
  end

  defp table(strings, v) do
    with {:ok, i} <- as_index(v),
         true <- i < tuple_size(strings) || :error do
      {:ok, elem(strings, i)}
    else
      _ -> {:error, :string_index_out_of_range}
    end
  end

  defp opt_table(_strings, nil), do: {:ok, nil}
  defp opt_table(strings, v), do: table(strings, v)

  defp as_index({:float, f}) when f >= 0 do
    i = trunc(f)
    if i * 1.0 == f, do: {:ok, i}, else: {:error, :non_integer_index}
  end

  defp as_index(_), do: {:error, :malformed_index}

  defp map_ok(items, fun) do
    Enum.reduce_while(items, {:ok, []}, fn item, {:ok, acc} ->
      case fun.(item) do
        {:ok, v} -> {:cont, {:ok, [v | acc]}}
        {:error, e} -> {:halt, {:error, e}}
      end
    end)
    |> case do
      {:ok, acc} -> {:ok, Enum.reverse(acc)}
      err -> err
    end
  end
end

File: implementations/haskell/src/Rhizomatic/Pack.hs
{-# LANGUAGE OverloadedStrings #-}

-- | The SPEC-8 L0 pack: a byte-deterministic physical container for delta
-- sets. Build factors shared envelope metadata out of transaction members;
-- unpack rehydrates and runs the §4 fsck — every record must rebuild to
-- canonical bytes whose multihash equals its stored id, or the unpack errors.
--
-- Layout lore pinned by the vectors and #19 F2–F4: every record carries "i"
-- (its own id hex as a strings index); the strings table sorts by RAW UTF-8
-- byte order, not the encoded-CBOR-key order (the length head dominates the
-- latter — two different orders, never unify them); version rides as a float
-- (the profile has no integer encoding).
module Rhizomatic.Pack (PackedDelta (..), buildPack, unpackPack, packIdHex) where

import qualified Data.ByteString as B
import Data.List (elemIndex, sortOn)
import qualified Data.Map.Strict as M
import Data.Maybe (fromMaybe, mapMaybe)
import qualified Data.Text as T
import qualified Data.Text.Encoding as TE
import Rhizomatic.Blake3 (blake3)
import Rhizomatic.Cbor (Item (..), decode, encode)
import Rhizomatic.Delta (Claims (..), Pointer (..), Target (..), deltaIdHex, validateClaims)
import Rhizomatic.Hex (encodeHex)

-- | A delta as it travels through a pack: claims plus optional signature hex.
data PackedDelta = PackedDelta {pdClaims :: Claims, pdSig :: Maybe T.Text} deriving (Eq, Show)

txnMember :: T.Text
txnMember = "rhizomatic.txn.member"

-- | SPEC-8 §3.1 (envelopes win, #19 F6): a delta is a manifest iff it carries
-- at least one pointer with role rhizomatic.txn.member AND a DeltaRef target.
isManifest :: Claims -> Bool
isManifest claims = not (null (memberRefs claims))

memberRefs :: Claims -> [T.Text]
memberRefs claims = [d | Pointer role (TDeltaRef d _) <- cPointers claims, role == txnMember]

-- Build -----------------------------------------------------------------------

buildPack :: [PackedDelta] -> Either String B.ByteString
buildPack deltas = do
  withIds0 <- mapM (\pd -> (,) pd <$> deltaIdHex (pdClaims pd)) deltas
  -- a delta set is a set: dedup by id (keep the first occurrence)
  let withIds = dedupOn snd withIds0
      envelopes = sortOn snd [x | x@(pd, _) <- withIds, isManifest (pdClaims pd)]
      envelopeIds = map snd envelopes
      -- member id -> index of the lexicographically first claiming manifest
      claimedBy =
        M.fromListWith min
          [ (T.unpack ref, mIdx)
            | (mIdx, (pd, _)) <- zip [0 :: Int ..] envelopes,
              ref <- memberRefs (pdClaims pd)
          ]
      members =
        sortOn snd
          [ x
            | x@(pd, idHex) <- withIds,
              not (isManifest (pdClaims pd)),
              M.member idHex claimedBy
          ]
      loose =
        sortOn snd
          [ x
            | x@(pd, idHex) <- withIds,
              not (isManifest (pdClaims pd)),
              not (M.member idHex claimedBy)
          ]
      envRecordOf (pd, idHex) = hydratedRecord pd idHex
      memberRecordOf (pd, idHex) =
        let mIdx = claimedBy M.! idHex
            (mPd, _) = envelopes !! mIdx
            mClaims = pdClaims mPd
            claims = pdClaims pd
            dt = cTimestamp claims - cTimestamp mClaims
         in ( [("i", SRef idHex), ("m", SNum (fromIntegral mIdx)), ("p", SPtrs (cPointers claims))]
                ++ [("a", SRef (T.unpack (cAuthor claims))) | cAuthor claims /= cAuthor mClaims]
                ++ [("dt", SNum dt) | dt /= 0]
                ++ [("s", SRef (T.unpack sig)) | Just sig <- [pdSig pd]]
            )
      envRecords = map envRecordOf envelopes
      memberRecords = map memberRecordOf members
      looseRecords = map envRecordOf loose
      -- the strings table: exactly the strings the records reference (#19 F4),
      -- sorted by raw UTF-8 byte order (#19 F3)
      allStrings =
        dedupOn id
          (concatMap recordStrings (envRecords ++ memberRecords ++ looseRecords))
      table = sortOn (TE.encodeUtf8 . T.pack) allStrings
      item =
        Map
          [ ("version", Num 1),
            ("strings", Arr (map (TStr . T.pack) table)),
            ("envelopes", Arr (map (recordToItem table) envRecords)),
            ("members", Arr (map (recordToItem table) memberRecords)),
            ("loose", Arr (map (recordToItem table) looseRecords))
          ]
  Right (encode item)

packIdHex :: B.ByteString -> String
packIdHex packBytes = encodeHex (B.pack [0x1e, 0x20] <> blake3 packBytes)

-- An abstract record field value, resolved against the strings table late so
-- the table can be collected from the records themselves.
data SVal = SRef String | SNum Double | SPtrs [Pointer]

hydratedRecord :: PackedDelta -> String -> [(T.Text, SVal)]
hydratedRecord pd idHex =
  let claims = pdClaims pd
   in [ ("a", SRef (T.unpack (cAuthor claims))),
        ("i", SRef idHex),
        ("t", SNum (cTimestamp claims)),
        ("p", SPtrs (cPointers claims))
      ]
        ++ [("s", SRef (T.unpack sig)) | Just sig <- [pdSig pd]]

recordStrings :: [(T.Text, SVal)] -> [String]
recordStrings = concatMap fieldStrings
  where
    fieldStrings (_, SRef s) = [s]
    fieldStrings (_, SNum _) = []
    fieldStrings (_, SPtrs ptrs) = concatMap ptrStrings ptrs
    ptrStrings (Pointer role target) =
      T.unpack role : case target of
        TEntity eid ctx -> T.unpack eid : ctxStrings ctx
        TDeltaRef dref ctx -> T.unpack dref : ctxStrings ctx
        TString s -> [T.unpack s]
        TNumber _ -> []
        TBool _ -> []
        TBytes mime _ -> [T.unpack mime]
    ctxStrings = maybe [] (\c -> [T.unpack c])

recordToItem :: [String] -> [(T.Text, SVal)] -> Item
recordToItem table fields = Map (map fieldToItem fields)
  where
    fieldToItem (key, val) = (key, valToItem val)
    valToItem (SRef s) = Num (idx s)
    valToItem (SNum n) = Num n
    valToItem (SPtrs ptrs) = Arr (map ptrToItem ptrs)
    ptrToItem (Pointer role target) =
      Map
        ( ("r", Num (idx (T.unpack role)))
            : case target of
              TEntity eid ctx -> ("e", Num (idx (T.unpack eid))) : ctxKv ctx
              TDeltaRef dref ctx -> ("d", Num (idx (T.unpack dref))) : ctxKv ctx
              TString s -> [("s", Num (idx (T.unpack s)))]
              TNumber n -> [("n", Num n)]
              TBool b -> [("b", Bool' b)]
              TBytes mime payload -> [("m", Num (idx (T.unpack mime))), ("y", BStr payload)]
        )
    ctxKv = maybe [] (\c -> [("c", Num (idx (T.unpack c)))])
    idx s = fromIntegral (fromMaybe (error ("pack: string not in table: " ++ s)) (elemIndex s table))

dedupOn :: Eq b => (a -> b) -> [a] -> [a]
dedupOn f = go []
  where
    go _ [] = []
    go seen (x : xs)
      | f x `elem` seen = go seen xs
      | otherwise = x : go (f x : seen) xs

-- Unpack ----------------------------------------------------------------------

-- | Decode pack bytes, rehydrate every record, and fsck: each rebuilt delta's
-- recomputed multihash MUST equal its stored id (SPEC-8 §4).
unpackPack :: B.ByteString -> Either String [PackedDelta]
unpackPack bytes = do
  item <- decode bytes
  top <- asMap "pack" item
  closedKeys' "pack" ["version", "strings", "envelopes", "members", "loose"] top
  version <- require "version" top
  case version of
    Num 1 -> Right ()
    _ -> Left "pack: unsupported version"
  table <- require "strings" top >>= asStrings
  envItems <- require "envelopes" top >>= asArr "envelopes"
  memberItems <- require "members" top >>= asArr "members"
  looseItems <- require "loose" top >>= asArr "loose"
  envelopes <- mapM (hydratedFromItem table) envItems
  members <- mapM (memberFromItem table envelopes) memberItems
  loose <- mapM (hydratedFromItem table) looseItems
  let everything = envelopes ++ members ++ loose
  mapM_ fsck everything
  Right (map fst everything)
  where
    fsck (pd, claimedId) = do
      validateClaims (pdClaims pd)
      idHex <- deltaIdHex (pdClaims pd)
      if idHex == claimedId
        then Right ()
        else Left ("pack: fsck failed — record id " ++ claimedId ++ " rehydrates to " ++ idHex)

hydratedFromItem :: [T.Text] -> Item -> Either String (PackedDelta, String)
hydratedFromItem table item = do
  fields <- asMap "record" item
  closedKeys' "record" ["a", "i", "t", "p", "s"] fields
  author <- require "a" fields >>= tableRef table
  idHex <- require "i" fields >>= tableRef table
  ts <- require "t" fields >>= asNum "t"
  ptrs <- require "p" fields >>= asArr "p" >>= mapM (ptrFromItem table)
  sig <- optionalRef table "s" fields
  Right (PackedDelta (Claims ts author ptrs) sig, T.unpack idHex)

memberFromItem :: [T.Text] -> [(PackedDelta, String)] -> Item -> Either String (PackedDelta, String)
memberFromItem table envelopes item = do
  fields <- asMap "member" item
  closedKeys' "member" ["m", "i", "p", "a", "dt", "s"] fields
  mIdxN <- require "m" fields >>= asNum "m"
  (envPd, _) <- indexInto "member envelope" envelopes mIdxN
  let mClaims = pdClaims envPd
  idHex <- require "i" fields >>= tableRef table
  ptrs <- require "p" fields >>= asArr "p" >>= mapM (ptrFromItem table)
  authorM <- optionalRef table "a" fields
  let author = fromMaybe (cAuthor mClaims) authorM
  dt <- case lookup "dt" fields of
    Nothing -> Right 0
    Just v -> asNum "dt" v
  sig <- optionalRef table "s" fields
  Right (PackedDelta (Claims (cTimestamp mClaims + dt) author ptrs) sig, T.unpack idHex)

ptrFromItem :: [T.Text] -> Item -> Either String Pointer
ptrFromItem table item = do
  fields <- asMap "ptr" item
  closedKeys' "ptr" ["r", "e", "d", "s", "n", "b", "m", "y", "c"] fields
  role <- require "r" fields >>= tableRef table
  ctx <- optionalRef table "c" fields
  let kinds = [k | k <- ["e", "d", "s", "n", "b", "y"], k `elem` map fst fields]
  target <- case kinds of
    ["e"] -> flip TEntity ctx <$> (require "e" fields >>= tableRef table)
    ["d"] -> flip TDeltaRef ctx <$> (require "d" fields >>= tableRef table)
    ["s"] -> do
      noContext ctx
      TString <$> (require "s" fields >>= tableRef table)
    ["n"] -> do
      noContext ctx
      TNumber <$> (require "n" fields >>= asNum "n")
    ["b"] -> do
      noContext ctx
      require "b" fields >>= \v -> case v of
        Bool' b -> Right (TBool b)
        _ -> Left "pack: b must be a bool"
    ["y"] -> do
      noContext ctx
      mime <- require "m" fields >>= tableRef table
      payload <- require "y" fields >>= \v -> case v of
        BStr bs -> Right bs
        _ -> Left "pack: y must be a byte string"
      Right (TBytes mime payload)
    _ -> Left "pack: ptr must carry exactly one target kind"
  Right (Pointer role target)
  where
    noContext Nothing = Right ()
    noContext (Just _) = Left "pack: a literal ptr carries no context"

-- helpers ---------------------------------------------------------------------

asMap :: String -> Item -> Either String [(T.Text, Item)]
asMap _ (Map kvs) = Right kvs
asMap what _ = Left ("pack: " ++ what ++ " must be a map")

asArr :: String -> Item -> Either String [Item]
asArr _ (Arr xs) = Right xs
asArr what _ = Left ("pack: " ++ what ++ " must be an array")

asNum :: String -> Item -> Either String Double
asNum _ (Num n) = Right n
asNum what _ = Left ("pack: " ++ what ++ " must be a number")

asStrings :: Item -> Either String [T.Text]
asStrings (Arr xs) = mapM (\x -> case x of TStr t -> Right t; _ -> Left "pack: strings must be text") xs
asStrings _ = Left "pack: strings must be an array"

require :: T.Text -> [(T.Text, Item)] -> Either String Item
require key fields = maybe (Left ("pack: missing key " ++ show key)) Right (lookup key fields)

optionalRef :: [T.Text] -> T.Text -> [(T.Text, Item)] -> Either String (Maybe T.Text)
optionalRef table key fields = case lookup key fields of
  Nothing -> Right Nothing
  Just v -> Just <$> tableRef table v

tableRef :: [T.Text] -> Item -> Either String T.Text
tableRef table (Num n) = snd <$> indexInto "strings" (map ((,) ()) table) n
tableRef _ _ = Left "pack: string index must be a number"

indexInto :: String -> [(a, b)] -> Double -> Either String (a, b)
indexInto what xs n =
  let i = truncate n :: Int
   in if fromIntegral i /= n || i < 0 || i >= length xs
        then Left ("pack: bad " ++ what ++ " index")
        else Right (xs !! i)

closedKeys' :: String -> [T.Text] -> [(T.Text, Item)] -> Either String ()
closedKeys' what allowed fields =
  case [k | (k, _) <- fields, k `notElem` allowed] of
    [] -> Right ()
    (k : _) -> Left ("pack: " ++ what ++ ": unknown key " ++ show k)
